File: medium/3254. Find the Power of K-Size Subarrays I/3254. Find the Power of K-Size Subarrays I.py
from typing import List


class Solution:
    def resultsArray(self, nums: List[int], k: int) -> List[int]:
        # res: List[int] = []

        # def run(nums: List[int]) -> None:
        #     nonlocal res

        #     if len(nums) < k:
        #         return

        #     sub_array: List[int] = nums[:k]
        #     is_power = self.check_power(sub_array)
        #     res.append(sub_array.pop() if is_power else -1)

        #     run(nums[1:])

        # run(nums)
        # return res

        if len(nums) < k:
            return

        sub_array: List[int] = nums[:k]
        is_power = self.check_power(sub_array)
        max_num = sub_array.pop() if is_power else -1
        next_num = self.resultsArray(nums[1:], k) or []

        return [max_num, *next_num]

    def check_power(self, nums: List[int]) -> bool:
        for i in range(0, len(nums) - 1, 1):
            if nums[i + 1] - nums[i] != 1:
                return False

        return True

File: medium/3254. Find the Power of K-Size Subarrays I/3254. Find the Power of K-Size Subarrays I.ts
function resultsArray(nums: number[], k: number): number[] {
    if (nums.length < k) return []

    function checkPower(start: number): boolean {
        for (let i = start; i < start + k - 1; i++) {
            if (nums[i + 1] - nums[i] != 1) return false
        }
        return true
    }

    function computeResults(index: number): number[] {
        if (index + k > nums.length) return []
        const isPower = checkPower(index)
        const maxNum = isPower ? nums[index + k - 1] : -1;
        return [maxNum, ...computeResults(index + 1)]
    }

    return computeResults(0)
};


// function check_power(nums: number[]): boolean {
//     for (let i = 0; i < nums.length; i++) {
//         if (nums[i + 1] - nums[i] != 1) return false
//     }
//     return true
// }

